File: src/list_stock.py
# list_stock.py
#
# Copyright 2025 Flávio de Vasconcellos Corrêa
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

import gi
import html
import locale

gi.require_version('Gtk', '4.0')
gi.require_version('Adw', '1')

from gi.repository import Gtk, Adw, GObject, Gio

from .stock import Stock


@Gtk.Template(resource_path='/com/github/sheepfarm/merkato/list_stock.ui')
class MerkatoListStock(Gtk.Box):
    __gtype_name__ = 'MerkatoListStock'

    _list_stock = Gtk.Template.Child()
    _empty_watchlist_state = Gtk.Template.Child()
    _list_scroll = Gtk.Template.Child()

    __gsignals__ = {
        'stock-selected': (GObject.SignalFlags.RUN_FIRST, None, (Stock,)),
        'stock-remove-requested': (GObject.SignalFlags.RUN_FIRST, None, (Stock,)),
        'empty-state-changed': (GObject.SignalFlags.RUN_FIRST, None, (bool,)),
    }

    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.remove_is_enabled = False
        self.current_sort = 'alphabetical'

        self.stock_list_store = Gio.ListStore.new(Stock)
        self._list_stock.bind_model(self.stock_list_store, self._create_stock_row)

        self._list_stock.connect('row-selected', self._on_row_selected)
        self.stock_list_store.connect("items-changed", self._on_items_changed)

        self._update_state()

    # ============== Métodos de ordenação ==============

    def sort_alphabetical(self):
        """Ordena alfabeticamente por nome."""
        self.current_sort = 'alphabetical'
        self._apply_sort()

    def sort_by_gains(self):
        """Ordena por maiores ganhos."""
        self.current_sort = 'gains'
        self._apply_sort()

    def sort_by_losses(self):
        """Ordena por maiores perdas."""
        self.current_sort = 'losses'
        self._apply_sort()

    def _apply_sort(self):
        """Aplica a ordenação atual."""
        stocks = self.get_all_stocks()

        if self.current_sort == 'alphabetical':
            stocks.sort(key=lambda s: s.long_name.lower())
        elif self.current_sort == 'gains':
            stocks.sort(
                key=lambda s: float(s.change_pct) if s.change_pct is not None else -float('inf'),
                reverse=True
            )
        elif self.current_sort == 'losses':
            stocks.sort(
                key=lambda s: float(s.change_pct) if s.change_pct is not None else float('inf')
            )

        self.stock_list_store.remove_all()
        for stock in stocks:
            self.stock_list_store.append(stock)

    # ============== Manipulação de stocks ==============

    def append(self, stock_item: Stock):
        """Adiciona um stock à lista."""
        if stock_item:
            self.stock_list_store.append(stock_item)

    def update(self, updated_stock: Stock) -> bool:
        """
        Atualiza um stock existente.

        Args:
            updated_stock: Stock com dados atualizados

        Returns:
            True se atualizado com sucesso
        """
        for i in range(self.stock_list_store.get_n_items()):
            stock = self.stock_list_store.get_item(i)
            if stock.symbol == updated_stock.symbol:
                stock.long_name = updated_stock.long_name
                stock.price = updated_stock.price
                stock.change = updated_stock.change
                stock.change_pct = updated_stock.change_pct
                stock.market_state = updated_stock.market_state
                stock.currency = updated_stock.currency
                stock.currency_symbol = updated_stock.currency_symbol

                self.stock_list_store.items_changed(i, 1, 1)

                if self.current_sort != 'alphabetical':
                    self._apply_sort()

                return True

        return False

    def remove_stock_by_symbol(self, symbol: str) -> bool:
        """
        Remove um stock pelo símbolo.

        Args:
            symbol: Símbolo do stock

        Returns:
            True se removido com sucesso
        """
        for i in range(self.stock_list_store.get_n_items()):
            stock = self.stock_list_store.get_item(i)
            if stock.symbol == symbol:
                self.stock_list_store.remove(i)
                return True
        return False

    def clear_all(self):
        """Remove todos os stocks."""
        self.stock_list_store.remove_all()

    def get_all_stocks(self) -> list:
        """Retorna lista com todos os stocks."""
        return [
            self.stock_list_store.get_item(i)
            for i in range(self.stock_list_store.get_n_items())
        ]

    def get_selected_stock(self) -> Stock:
        """Retorna o stock selecionado."""
        selected_row = self._list_stock.get_selected_row()
        return selected_row.stock_item if selected_row else None

    def is_empty(self) -> bool:
        """Verifica se a lista está vazia."""
        return self.stock_list_store.get_n_items() == 0

    # ============== Modo de remoção ==============

    def set_remove_enabled(self, enabled: bool):
        """
        Habilita/desabilita o modo de remoção.

        Args:
            enabled: True para habilitar
        """
        self.remove_is_enabled = enabled

        for i in range(self.stock_list_store.get_n_items()):
            row = self._list_stock.get_row_at_index(i)
            if row and hasattr(row, 'remove_button'):
                if enabled:
                    row.remove_button.add_css_class("enabled")
                else:
                    row.remove_button.remove_css_class("enabled")

    # ============== Criação de linhas ==============

    def _create_stock_row(self, stock_item: Stock) -> Gtk.ListBoxRow:
        """
        Cria uma linha para exibir um stock.

        Args:
            stock_item: Stock a ser exibido

        Returns:
            Widget da linha
        """
        if not stock_item:
            return None

        row = Adw.ActionRow()
        row.set_activatable(True)
        row.set_cursor_from_name("pointer")
        row.stock_item = stock_item

        # Título e subtítulo
        if stock_item.long_name:
            row.set_title(html.escape(stock_item.long_name))
        if stock_item.symbol:
            row.set_subtitle(stock_item.symbol)

        # Box de informações (preço e variação)
        suffix_box = self._create_price_box(stock_item)
        row.add_suffix(suffix_box)

        # Botão de remoção
        remove_button = self._create_remove_button(stock_item)
        row.remove_button = remove_button
        row.add_suffix(remove_button)

        # Estilo baseado no estado do mercado
        self._apply_market_state_style(row, stock_item)

        # Ação ao clicar (abre no Yahoo Finance)
        url = f"https://finance.yahoo.com/quote/{stock_item.symbol}/"
        row.connect("activated", lambda r: Gio.AppInfo.launch_default_for_uri(url, None))

        return row

    def _create_price_box(self, stock_item: Stock) -> Gtk.Box:
        """Cria o box com preço e variação."""
        suffix_box = Gtk.Box(orientation=Gtk.Orientation.VERTICAL)
        suffix_box.set_halign(Gtk.Align.END)
        suffix_box.set_valign(Gtk.Align.CENTER)
        suffix_box.set_spacing(2)

        # Label de preço
        price_label = Gtk.Label()
        price_label.set_halign(Gtk.Align.END)
        price_label.add_css_class("numeric")

        currency_symbol = '' if stock_item.symbol.startswith("^") else stock_item.currency
        price_fmt = locale.format_string("%.2f", stock_item.price, grouping=True, monetary=True)
        price_label.set_label(f"{price_fmt} {currency_symbol}")

        # Label de variação
        change_label = Gtk.Label()
        change_label.set_halign(Gtk.Align.END)
        change_label.add_css_class("caption")

        change_fmt = locale.format_string("%.2f", stock_item.change, grouping=True, monetary=True)
        change_pct_fmt = locale.format_string("%.2f", stock_item.change_pct * 100, grouping=True, monetary=True)

        if stock_item.change >= 0:
            change_label.set_label(f"+{change_fmt} ({change_pct_fmt}%)")
            change_label.add_css_class("success")
        else:
            change_label.set_label(f"{change_fmt} ({change_pct_fmt}%)")
            change_label.add_css_class("error")

        suffix_box.append(price_label)
        suffix_box.append(change_label)

        return suffix_box

    def _create_remove_button(self, stock_item: Stock) -> Gtk.Button:
        """Cria o botão de remoção."""
        remove_button = Gtk.Button()
        remove_button.set_icon_name("user-trash-symbolic")
        remove_button.set_valign(Gtk.Align.CENTER)
        remove_button.add_css_class("flat")
        remove_button.add_css_class("circular")
        remove_button.add_css_class("remove-button")
        remove_button.set_tooltip_text(_("Remove from watchlist"))
        remove_button.connect("clicked", self._on_remove_button_clicked, stock_item)

        if self.remove_is_enabled:
            remove_button.add_css_class("enabled")

        return remove_button

    def _apply_market_state_style(self, row: Gtk.Widget, stock_item: Stock):
        """Aplica estilo baseado no estado do mercado."""
        if stock_item.market_state == "REGULAR":
            row.remove_css_class("market-closed")
            row.add_css_class("market-opened")
        else:
            row.remove_css_class("market-opened")
            row.add_css_class("market-closed")

    # ============== Callbacks ==============

    def _on_row_selected(self, listbox, row):
        """Callback quando uma linha é selecionada."""
        if row:
            self.emit('stock-selected', row.stock_item)

    def _on_remove_button_clicked(self, button, stock_item):
        """Callback quando o botão de remoção é clicado."""
        self.emit('stock-remove-requested', stock_item)

    def _on_items_changed(self, list_store, position, removed, added):
        """Callback quando itens são modificados na lista."""
        self._update_state()

    def _update_state(self):
        """Atualiza o estado de visibilidade (vazio/com conteúdo)."""
        is_empty = self.is_empty()
        has_items = not is_empty

        self._list_scroll.set_visible(has_items)
        self._empty_watchlist_state.set_visible(is_empty)

        self.emit('empty-state-changed', is_empty)

File: src/list_stock.rs
// list_stock.rs
//
// Copyright 2025 Flávio de Vasconcellos Corrêa
//
// SPDX-License-Identifier: GPL-3.0-or-later

use std::cell::RefCell;

use gtk4::prelude::*;
use gtk4::subclass::prelude::*;
use gtk4::{gio, glib, CompositeTemplate};
use libadwaita::prelude::{ActionRowExt, PreferencesRowExt};

use crate::stock::Stock;
use crate::stock_object::StockObject;

// ─── Sort mode ────────────────────────────────────────────────────────────────

#[derive(Debug, Clone, PartialEq)]
pub enum SortMode {
    Alphabetical,
    TopGains,
    TopLosses,
}

impl SortMode {
    pub fn from_str(s: &str) -> Self {
        match s {
            "gains" => SortMode::TopGains,
            "losses" => SortMode::TopLosses,
            _ => SortMode::Alphabetical,
        }
    }

    #[allow(dead_code)]
    pub fn as_str(&self) -> &'static str {
        match self {
            SortMode::Alphabetical => "alphabetical",
            SortMode::TopGains => "gains",
            SortMode::TopLosses => "losses",
        }
    }
}

// ─── GObject subclass ─────────────────────────────────────────────────────────

mod imp {
    use super::*;
    use gtk4::TemplateChild;

    #[derive(CompositeTemplate)]
    #[template(resource = "/com/github/sheepfarm/merkato/list_stock.ui")]
    pub struct ListStock {
        #[template_child(id = "_list_scroll")]
        pub list_scroll: TemplateChild<gtk4::ScrolledWindow>,
        #[template_child(id = "_list_stock")]
        pub list_view: TemplateChild<gtk4::ListView>,
        #[template_child(id = "_empty_watchlist_state")]
        pub empty_state: TemplateChild<libadwaita::StatusPage>,

        pub model: RefCell<Option<gio::ListStore>>,
        pub sort_mode: RefCell<SortMode>,
        pub remove_mode: RefCell<bool>,
        pub remove_callback: RefCell<Option<Box<dyn Fn(String)>>>,
        pub add_alert_callback: RefCell<Option<Box<dyn Fn(String)>>>,
    }

    impl Default for ListStock {
        fn default() -> Self {
            Self {
                list_scroll: Default::default(),
                list_view: Default::default(),
                empty_state: Default::default(),
                model: RefCell::new(None),
                sort_mode: RefCell::new(SortMode::Alphabetical),
                remove_mode: RefCell::new(false),
                remove_callback: RefCell::new(None),
                add_alert_callback: RefCell::new(None),
            }
        }
    }

    impl std::fmt::Debug for ListStock {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            f.debug_struct("ListStock").finish()
        }
    }

    #[glib::object_subclass]
    impl ObjectSubclass for ListStock {
        const NAME: &'static str = "MerkatoListStock";
        type Type = super::ListStock;
        type ParentType = gtk4::Box;

        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
        }
        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }
    }

    impl ObjectImpl for ListStock {
        fn constructed(&self) {
            self.parent_constructed();
            let widget = self.obj();
            widget.setup_factory();
        }
    }

    impl WidgetImpl for ListStock {}
    impl BoxImpl for ListStock {}
}

// ─── Public wrapper ───────────────────────────────────────────────────────────

glib::wrapper! {
    pub struct ListStock(ObjectSubclass<imp::ListStock>)
        @extends gtk4::Box, gtk4::Widget,
        @implements gtk4::Buildable, gtk4::ConstraintTarget, gtk4::Orientable;
}

impl ListStock {
    pub fn new() -> Self {
        glib::Object::new()
    }

    fn setup_factory(&self) {
        let imp = self.imp();
        let model = gio::ListStore::new::<StockObject>();
        imp.list_view.set_model(Some(&gtk4::NoSelection::new(Some(model.clone()))));
        *imp.model.borrow_mut() = Some(model);

        let factory = gtk4::SignalListItemFactory::new();
        factory.connect_bind(glib::clone!(
            #[weak(rename_to = list_widget)]
            self,
            move |_, list_item| {
                let list_item = list_item.downcast_ref::<gtk4::ListItem>().unwrap();
                list_item.set_child(None::<&gtk4::Widget>);

                let Some(obj) = list_item.item().and_downcast::<StockObject>() else { return };
                let stock = obj.stock();
                let row = list_widget.build_row(&stock);
                list_item.set_child(Some(&row));
            }
        ));

        imp.list_view.set_factory(Some(&factory));
    }

    fn build_row(&self, stock: &Stock) -> libadwaita::ActionRow {
        let imp = self.imp();
        let remove_mode = *imp.remove_mode.borrow();

        let row = libadwaita::ActionRow::new();
        row.set_title(&glib::markup_escape_text(&stock.long_name));
        row.set_subtitle(&stock.symbol);
        row.set_activatable(true);
        row.set_cursor_from_name(Some("pointer"));

        if stock.is_market_open() {
            row.add_css_class("market-opened");
        } else {
            row.add_css_class("market-closed");
        }

        let symbol = stock.symbol.clone();
        row.connect_activated(move |_| {
            let url = format!("https://finance.yahoo.com/quote/{}/", symbol);
            let _ = gio::AppInfo::launch_default_for_uri(&url, gio::AppLaunchContext::NONE);
        });

        row.add_suffix(&self.build_price_box(stock));
        row.add_suffix(&self.build_remove_button(&stock.symbol, remove_mode));
        self.attach_context_menu(&row, stock);

        row
    }

    fn build_price_box(&self, stock: &Stock) -> gtk4::Box {
        let vbox = gtk4::Box::new(gtk4::Orientation::Vertical, 2);
        vbox.set_halign(gtk4::Align::End);
        vbox.set_valign(gtk4::Align::Center);

        let price_str = if stock.symbol.starts_with('^') {
            format!("{:.2}", stock.price)
        } else {
            format!("{:.2} {}", stock.price, stock.currency_symbol)
        };
        let price_label = gtk4::Label::new(Some(&price_str));
        price_label.set_halign(gtk4::Align::End);
        price_label.add_css_class("numeric");

        let sign = if stock.change >= 0.0 { "+" } else { "" };
        let change_str = format!("{}{:.2} ({:.2}%)", sign, stock.change, stock.change_pct * 100.0);
        let change_label = gtk4::Label::new(Some(&change_str));
        change_label.set_halign(gtk4::Align::End);
        change_label.add_css_class("caption");
        change_label.add_css_class("numeric");

        if stock.change_pct > 0.0 {
            change_label.add_css_class("success");
        } else if stock.change_pct < 0.0 {
            change_label.add_css_class("error");
        }

        vbox.append(&price_label);
        vbox.append(&change_label);
        vbox
    }

    fn build_remove_button(&self, symbol: &str, visible: bool) -> gtk4::Button {
        let btn = gtk4::Button::new();
        btn.set_icon_name("user-trash-symbolic");
        btn.add_css_class("destructive-action");
        btn.add_css_class("circular");
        btn.add_css_class("flat");
        btn.set_valign(gtk4::Align::Center);
        btn.set_tooltip_text(Some("Remove from watchlist"));
        btn.set_visible(visible);

        let sym = symbol.to_string();
        btn.connect_clicked(glib::clone!(
            #[weak(rename_to = list_widget)]
            self,
            move |_| {
                if let Some(cb) = list_widget.imp().remove_callback.borrow().as_ref() {
                    cb(sym.clone());
                }
            }
        ));
        btn
    }

    fn attach_context_menu(&self, row: &libadwaita::ActionRow, stock: &Stock) {
        let menu = gio::Menu::new();
        menu.append(Some("Add Price Alert"), Some("row.add-alert"));
        menu.append(Some("Open in Yahoo Finance"), Some("row.open-yahoo"));

        let popover = gtk4::PopoverMenu::from_model(Some(&menu));
        popover.set_parent(row);
        popover.set_has_arrow(false);

        // Unparent the popover before the row is finalized to avoid warnings
        let popover_for_destroy = popover.clone();
        row.connect_destroy(move |_| {
            popover_for_destroy.unparent();
        });

        let action_group = gio::SimpleActionGroup::new();

        let add_alert = gio::SimpleAction::new("add-alert", None);
        let sym = stock.symbol.clone();
        add_alert.connect_activate(glib::clone!(
            #[weak(rename_to = list_widget)]
            self,
            move |_, _| {
                if let Some(cb) = list_widget.imp().add_alert_callback.borrow().as_ref() {
                    cb(sym.clone());
                }
            }
        ));
        action_group.add_action(&add_alert);

        let open_yahoo = gio::SimpleAction::new("open-yahoo", None);
        let url = format!("https://finance.yahoo.com/quote/{}/", stock.symbol);
        open_yahoo.connect_activate(move |_, _| {
            let _ = gio::AppInfo::launch_default_for_uri(&url, gio::AppLaunchContext::NONE);
        });
        action_group.add_action(&open_yahoo);

        row.insert_action_group("row", Some(&action_group));

        let gesture = gtk4::GestureClick::new();
        gesture.set_button(3);
        let popover_ref = popover.clone();
        gesture.connect_pressed(move |gesture, _, x, y| {
            let rect = gdk4::Rectangle::new(x as i32, y as i32, 1, 1);
            popover_ref.set_pointing_to(Some(&rect));
            popover_ref.popup();
            gesture.set_state(gtk4::EventSequenceState::Claimed);
        });
        row.add_controller(gesture);
    }

    // ── Data ─────────────────────────────────────────────────────────────────

    pub fn set_stocks(&self, stocks: Vec<Stock>) {
        let sorted = self.sorted_stocks(stocks);
        let model = gio::ListStore::new::<StockObject>();
        for stock in sorted {
            model.append(&StockObject::new(stock));
        }
        self.imp().list_view.set_model(Some(&gtk4::NoSelection::new(Some(model.clone()))));
        *self.imp().model.borrow_mut() = Some(model);

        self.update_empty_state();
    }

    pub fn set_sort_mode(&self, mode: SortMode) {
        *self.imp().sort_mode.borrow_mut() = mode;
        if let Some(model) = self.imp().model.borrow().clone() {
            let mut stocks: Vec<Stock> = (0..model.n_items())
                .filter_map(|i| model.item(i).and_downcast::<StockObject>().map(|o| o.stock()))
                .collect();
            stocks = self.sorted_stocks(stocks);
            model.remove_all();
            for stock in stocks {
                model.append(&StockObject::new(stock));
            }
        }
    }

    pub fn set_remove_mode(&self, enabled: bool) {
        *self.imp().remove_mode.borrow_mut() = enabled;
        self.refresh_all_rows();
    }

    pub fn sort_mode(&self) -> SortMode {
        self.imp().sort_mode.borrow().clone()
    }

    fn sorted_stocks(&self, mut stocks: Vec<Stock>) -> Vec<Stock> {
        match *self.imp().sort_mode.borrow() {
            SortMode::Alphabetical => {
                stocks.sort_by(|a, b| a.long_name.to_lowercase().cmp(&b.long_name.to_lowercase()));
            }
            SortMode::TopGains => {
                stocks.sort_by(|a, b| {
                    b.change_pct
                        .partial_cmp(&a.change_pct)
                        .unwrap_or(std::cmp::Ordering::Equal)
                });
            }
            SortMode::TopLosses => {
                stocks.sort_by(|a, b| {
                    a.change_pct
                        .partial_cmp(&b.change_pct)
                        .unwrap_or(std::cmp::Ordering::Equal)
                });
            }
        }
        stocks
    }

    fn update_empty_state(&self) {
        let imp = self.imp();
        let count = imp.model.borrow().as_ref().map(|m| m.n_items()).unwrap_or(0);
        imp.list_scroll.set_visible(count > 0);
        imp.empty_state.set_visible(count == 0);
    }

    fn refresh_all_rows(&self) {
        let imp = self.imp();
        if let Some(model) = imp.model.borrow().clone() {
            // Replace objects in reverse order so indices remain valid
            let mut i = model.n_items();
            while i > 0 {
                i -= 1;
                if let Some(obj) = model.item(i).and_downcast::<StockObject>() {
                    let stock = obj.stock();
                    model.remove(i);
                    model.insert(i, &StockObject::new(stock));
                }
            }
        }
    }

    // ── Callbacks ─────────────────────────────────────────────────────────────

    pub fn connect_remove_requested<F: Fn(String) + 'static>(&self, f: F) {
        *self.imp().remove_callback.borrow_mut() = Some(Box::new(f));
    }

    pub fn connect_add_alert<F: Fn(String) + 'static>(&self, f: F) {
        *self.imp().add_alert_callback.borrow_mut() = Some(Box::new(f));
    }
}

impl Default for ListStock {
    fn default() -> Self {
        Self::new()
    }
}
